File: nodejs/lib/varint.js
'use strict';

/**
 * unsigned varint (uvarint) encode/decode
 * - Same LEB128 encoding as protobuf
 * - Small indices use 1 byte, large indices are also efficiently encoded
 */

/**
 * Encode uint32 value as uvarint, return Buffer
 * @param {number} value - Non-negative integer
 * @returns {Buffer}
 */
function encode(value) {
  const bytes = [];
  while (value > 0x7F) {
    bytes.push((value & 0x7F) | 0x80);
    value >>>= 7;
  }
  bytes.push(value & 0x7F);
  return Buffer.from(bytes);
}

/**
 * Encode multiple uint32 values as consecutive uvarints
 * @param {number[]} values
 * @returns {Buffer}
 */
function encodeArray(values) {
  const parts = [];
  for (const v of values) {
    parts.push(encode(v));
  }
  return Buffer.concat(parts);
}

/**
 * Decode uvarint at offset in Buffer
 * @param {Buffer} buf
 * @param {number} offset
 * @returns {{ value: number, bytesRead: number }}
 */
function decode(buf, offset) {
  let value = 0;
  let shift = 0;
  let bytesRead = 0;

  while (offset < buf.length) {
    const b = buf[offset];
    value |= (b & 0x7F) << shift;
    offset++;
    bytesRead++;
    if ((b & 0x80) === 0) {
      return { value: value >>> 0, bytesRead };
    }
    shift += 7;
    if (shift > 35) {
      throw new Error('varint overflow: value too large');
    }
  }

  throw new Error('varint incomplete: insufficient data');
}

/**
 * Decode N consecutive uvarints from Buffer
 * @param {Buffer} buf
 * @param {number} offset
 * @param {number} count
 * @returns {{ values: number[], bytesRead: number }}
 */
function decodeArray(buf, offset, count) {
  const values = [];
  let totalRead = 0;

  for (let i = 0; i < count; i++) {
    const { value, bytesRead } = decode(buf, offset + totalRead);
    values.push(value);
    totalRead += bytesRead;
  }

  return { values, bytesRead: totalRead };
}

module.exports = { encode, encodeArray, decode, decodeArray };

File: python/src/mshzip/varint.py
'unsigned varint (uvarint) encode/decode - protobuf LEB128 compatible'
from __future__ import annotations


def encode(value: int) -> bytes:
    'Encode uint32 value as uvarint, return bytes.'
    result = bytearray()
    while value > 0x7F:
        result.append((value & 0x7F) | 0x80)
        value >>= 7
    result.append(value & 0x7F)
    return bytes(result)


def encode_array(values: list[int]) -> bytes:
    'Encode multiple uint32 values as consecutive uvarints.'
    parts = bytearray()
    for v in values:
        parts.extend(encode(v))
    return bytes(parts)


def decode(buf: bytes | bytearray | memoryview, offset: int = 0) -> tuple[int, int]:
    'Returns (value, bytes_read).'
    value = 0
    shift = 0
    bytes_read = 0
    pos = offset

    while pos < len(buf):
        b = buf[pos]
        value |= (b & 0x7F) << shift
        pos += 1
        bytes_read += 1
        if (b & 0x80) == 0:
            return (value & 0xFFFFFFFF, bytes_read)
        shift += 7
        if shift > 35:
            raise ValueError('varint overflow: value too large')

    raise ValueError('varint incomplete: insufficient data')


def decode_array(
    buf: bytes | bytearray | memoryview,
    offset: int,
    count: int,
) -> tuple[list[int], int]:
    'Returns (values, total_bytes_read).'
    values: list[int] = []
    total_read = 0

    for _ in range(count):
        value, bytes_read = decode(buf, offset + total_read)
        values.append(value)
        total_read += bytes_read

    return (values, total_read)
